initramfs/src/image.rs

```diff
@@ -0,0 +1,172 @@
+use std::io::Cursor;
+
+use anyhow::{anyhow, Result};
+use bytes::Bytes;
+use flate2::read::GzDecoder;
+use log::{debug, error, info};
+use serde::Deserialize;
+use tar::Archive;
+
+#[derive(Deserialize, Debug)]
+#[allow(non_snake_case)]
+struct Layer {
+    blobSum: String,
+}
+
+#[derive(Deserialize)]
+#[allow(non_snake_case)]
+struct ManifestResponse {
+    fsLayers: Vec<Layer>,
+}
+#[derive(Deserialize)]
+struct TokenResponse {
+    token: String,
+}
+
+#[derive(Debug)]
+pub struct Image {
+    pub registry: String,
+    pub tag: String,
+    pub repository: String,
+    pub auth_api: String,
+    token: String,
+    layers: Vec<Layer>,
+}
+
+impl Image {
+    pub fn new(image: &str, repository: String, auth_api: String) -> Result<Self> {
+        // Extract the registry and tag from the image
+        let mut parts = image.split(':');
+        let registry = parts.next().ok_or_else(|| anyhow!("Invalid image"))?;
+        let tag = parts.next().unwrap_or("latest");
+
+        // If the registry doesn't contain a '/', it's in "library/", meaning it's from the docker hub official images
+        let image = Self {
+            registry: if !registry.contains('/') {
+                format!("library/{}", registry)
+            } else {
+                registry.to_string()
+            },
+            tag: tag.to_string(),
+            repository,
+            auth_api,
+            token: "".to_string(),
+            layers: Vec::new(),
+        };
+
+        debug!("Successfully parsed image : {:?}", image);
+
+        Ok(image)
+    }
+
+    async fn get_auth_token(&mut self) -> Result<()> {
```

Comment:
nitty nit: I'd call that one `fetch_auth_token`to avoid confusing it with a getter.

Reply:
Ok. I changed it in the last force push !